File: apps/web/lib/deployments.ts
/**
 * Server-only helper for reading contract deployment addresses. Ticket S2-03.
 *
 * `deployments/<network>/addresses.json` is written by `contracts/scripts/deploy.ts`
 * (`deployments/localhost/` is gitignored and regenerated on every chain restart;
 * `deployments/amoy/` is committed so the web app can read it client-side-adjacent, i.e. here,
 * server-side, before handing the address down to the browser). On a fresh checkout nothing
 * has been deployed yet, so "the file doesn't exist" is an expected, non-error state — every
 * function here returns `null` rather than throwing.
 */

import { readFile } from "node:fs/promises";
import { join } from "node:path";

export type NetworkName = "localhost" | "amoy";

interface DeploymentAddresses {
  network: string;
  chainId: number;
  deployedAt: string;
  deployer: string;
  contracts: {
    ActorRegistry: string;
    DeviceRegistry: string;
    LotRegistry: string;
    BatchAnchor: string;
  };
}

/**
 * Which `deployments/` subdirectory to read for a given chain id. Mirrors
 * `contracts/hardhat.config.ts`: 31337 is the local Hardhat node, anything else is treated as
 * Amoy (80002). Deliberately derived from `NEXT_PUBLIC_VERIFY_CHAIN_ID` rather than a second
 * env var, so the network and the chain id can never disagree.
 */
export function networkForChainId(chainId: number): NetworkName {
  return chainId === 31337 ? "localhost" : "amoy";
}

/**
 * Read `deployments/<network>/addresses.json` relative to the repo root (from `apps/web`,
 * that's `../../deployments/<network>/addresses.json`). Returns `null` — never throws — when
 * the file is missing, which is the normal state before `npm run deploy:local` /
 * `npm run deploy:amoy` has been run.
 */
export async function readDeployment(network: NetworkName): Promise<DeploymentAddresses | null> {
  try {
    const path = join(process.cwd(), "..", "..", "deployments", network, "addresses.json");
    const raw = await readFile(path, "utf8");
    return JSON.parse(raw) as DeploymentAddresses;
  } catch {
    return null;
  }
}

/** Convenience: just the `BatchAnchor` address the verify page needs, or `null`. */
export async function readBatchAnchorAddress(
  network: NetworkName,
): Promise<`0x${string}` | null> {
  const deployment = await readDeployment(network);
  const address = deployment?.contracts.BatchAnchor;
  return (address as `0x${string}` | undefined) ?? null;
}

File: apps/android/app/src/main/kotlin/com/krishichain/app/domain/Deployments.kt
package com.krishichain.app.domain

import android.content.Context
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.io.IOException

/**
 * Kotlin port of `apps/web/lib/deployments.ts`. That file reads `deployments/<network>/addresses.json`
 * straight off disk because it runs server-side in Next.js; a mobile app has no repo checkout to
 * read at runtime, so the same JSON shape is bundled as an Android asset instead
 * (`assets/deployments/<network>/addresses.json`) — a placeholder/absent file today, since
 * nothing is deployed to any network in this repo yet (mirrors CLAUDE.md's "no fake verified
 * state" invariant: absence here must render honestly, not silently).
 *
 * On top of the asset path, `BuildConfig.BATCH_ANCHOR_ADDRESS_OVERRIDE` (from `gradle.properties`,
 * see that file) lets a build be pointed at a real deployment without needing an app rebuild-time
 * asset refresh for every `npm run deploy:*` — the override wins when non-blank; otherwise this
 * falls back to the bundled asset, and finally to `null` ("not deployed").
 */

enum class NetworkName(val raw: String) {
    LOCALHOST("localhost"),
    AMOY("amoy"),
}

/** Mirrors `contracts/hardhat.config.ts`: 31337 is the local Hardhat node, anything else is
 *  treated as Amoy (80002). Derived from the configured chain id so network and chain id can
 *  never disagree. */
fun networkForChainId(chainId: Int): NetworkName =
    if (chainId == 31337) NetworkName.LOCALHOST else NetworkName.AMOY

@Serializable
data class DeploymentContracts(
    val ActorRegistry: String? = null,
    val DeviceRegistry: String? = null,
    val LotRegistry: String? = null,
    val BatchAnchor: String? = null,
)

@Serializable
data class DeploymentAddresses(
    val network: String? = null,
    val chainId: Int? = null,
    val deployedAt: String? = null,
    val deployer: String? = null,
    val contracts: DeploymentContracts? = null,
)

private val json = Json { ignoreUnknownKeys = true }

/** Read `assets/deployments/<network>/addresses.json`. Returns `null` — never throws — when the
 *  asset is missing or malformed, which is the normal state before a deployment exists. */
fun readDeployment(context: Context, network: NetworkName): DeploymentAddresses? {
    return try {
        val raw = context.assets.open("deployments/${network.raw}/addresses.json")
            .bufferedReader(Charsets.UTF_8)
            .use { it.readText() }
        json.decodeFromString<DeploymentAddresses>(raw)
    } catch (_: IOException) {
        null
    } catch (_: Exception) {
        null
    }
}

/** Convenience: just the `BatchAnchor` address the verify screen needs, or `null` when nothing
 *  is configured or deployed for this network. */
fun readBatchAnchorAddress(context: Context, network: NetworkName, override: String): Hex? {
    if (override.isNotBlank()) return override
    val address = readDeployment(context, network)?.contracts?.BatchAnchor
    return address?.takeIf { it.isNotBlank() }
}
